deployment/secure_research_environment/cloud_init/resources/guacamole_docker_compose.mustache.yaml

```diff
@@ -26,11 +26,16 @@ services:
             POSTGRES_DB: guacamole
             POSTGRES_USER: guacamole
             POSTGRES_PASSWORD: {{{guacamole.internalDbPassword}}}
+        healthcheck:
```

Comment:
Deploying guacamole failed twice after I'd updated to the newer version, seemingly because the Postgres database had not finished initializing when docker started Guacamole. After adding the healthcheck it's worked fine. Not sure if there's a way to find logs of the healthcheck.

Reply:
`HEALTHCHECK` only adds a "health" status for containers. It doesn't restart them if they fail, you need another service to do that.

([I wrote one](https://github.com/JimMadge/healthy) for my homeserver :smile:)

Reply:
Which is fine - all I want it to do is make guacamole wait until postgres is finished initialising and healthy, not restart anything.

Reply:
that's why I did this, it wasn't waiting.

Reply:
I would be surprised if this fixes that problem. I think it is a race condition.

If you look at the cloud-init `docker-compose up -d` is run before the database is initialised.
It might be better to add a restart after doing that, or not bringing the Guacamole/Guacd container up until after initialising the database.

Reply:
I think I have indeed slightly misdiagnosed the problem, but think the healthcheck may actually fix it.

Here's part of the cloud-init log from a failed deployment, without the healthcheck implemented.

![image](https://github.com/alan-turing-institute/data-safe-haven/assets/5796417/5153d5f8-4114-4372-a811-a2f792d2ed2f)

The line of code that fails is 

`su guacamoledaemon -c "docker compose -f /opt/guacamole/docker-compose.yaml exec -T postgres psql -U guacamole -f /scripts/db_init.sql"`

The previous errors I've seen have been slightly different (always at the same point), but the general issue is the same - the database isn't ready to run the `db_init.sql` script.

So it's not that guacamole isn't up - it's a race, as you say. 

However, I think the healthcheck prevents this from happening, because it forces guacamole to wait until the db is healthy and accepting connections before starting. So by extension, by the time we get to the line that runs `db_init.sql`, the db is ready to run it, and we don't hit this problem.